ruby_programming/basic_ruby/nested_collections.md

```diff
@@ -0,0 +1,330 @@
+vehicles
+#=> {:alice=>{:year=>2019, :make=>"Toyota", :model=>"Corolla"}, :blake=>{:year=>2020, :make=>"Volkswagen", :model=>"Beetle"}, :caleb=>{:year=>2020, :make=>"Honda", :model=>"Accord"}, :dave=>{:year=>2021, :make=>"Ford", :model=>"Escape", :color=>"red"}}
+~~~
+
+Deleting one of the nested hashes will be just like a regular hash. Let's say Blake has decided to sell his Beetle and backpack across Europe. So, let's delete Blake's car.
+
+~~~ruby
+vehicles.delete(:blake)
+#=> {:year=>2020, :make=>"Volkswagen", :model=>"Beetle"}
+vehicles
+#=> {:alice=>{:year=>2019, :make=>"Toyota", :model=>"Corolla"}, :caleb=>{:year=>2020, :make=>"Honda", :model=>"Accord"}, :dave=>{:year=>2021, :make=>"Ford", :model=>"Escape", :color=>"red"}}
+~~~
+
+To delete one of the key/value pairs inside of a nested hash, you first specify the key of the hash. Let's say Dave decided that we don't need to specify the color of his vehicle. Therefore, we specify `vehicles[:dave]` before we indicate the key of the nested hash to delete.
+
+~~~ruby
+vehicles[:dave].delete(:color)
+#=> "red"
+vehicles
+#=> {:alice=>{:year=>2019, :make=>"Toyota", :model=>"Corolla"}, :caleb=>{:year=>2020, :make=>"Honda", :model=>"Accord"}, :dave=>{:year=>2021, :make=>"Ford", :model=>"Escape"}}
+~~~
+
+### Methods
+There are many helpful methods to use with nested hashes. Once you know what data you need from a nested hash, I have found that browsing through the documentation and experimenting with them in IRB is the best way for me to understand how they work.
+
+Let's look at an example using the vehicles nested hash. Let's say that we want to know who owns vehicles that are from 2020 or newer. At first glance in the documentation, it looks like `#select` would be a great method to use.
+
+~~~ruby
+vehicles.select { |_name, data| data[:year] >= 2020 }
+#=> {:caleb=>{:year=>2020, :make=>"Honda", :model=>"Accord"}, :dave=>{:year=>2021, :make=>"Ford", :model=>"Escape"}}
+~~~
+
+Yes, using `#select` gives us the information that we need. However, what if we only want the names of the owners and not another nested hash. Let's keep looking through the documentation to see what else we find. The `#collect` method sounds very useful for this situation. Let's collect the names if the year is 2020 or newer.
+
+~~~ruby
+vehicles.collect { |name, data| name if data[:year] >= 2020 }
+#=> [nil, :caleb, :dave]
+~~~
+
+Using `#collect` gets us a lot closer to only having the names of the new vehicle owners. If you look at this method in the documentation you will see that `#collect` and `#map` have the same functionality. Both of these methods use the return value of each interation, so when the if statement is false, it will return a nil value.
+
+Nil values can cause problems down the road, so let's look through the documentation to see if we can find a method to help solve this problem. The `#compact` method returns an array (or hash) without nil values, so let's chain it on the end of the block.
```

Comment:
🔥  brilliant irl example

Reply:
Thank you! Austin suggested adding `#filter_map` to this section too, so I am updating this section.